File: tokyo-x/src/voice/stt.ts
import { optionalEnv, providerKey } from "../core/env.js";

export interface TranscriptResult {
  ok: boolean;
  text: string;
  provider: string;
  reason?: string;
}

export interface Transcriber {
  readonly name: string;
  transcribe(audio: Buffer, mimeType: string): Promise<TranscriptResult>;
}

export class PlaceholderTranscriber implements Transcriber {
  readonly name = "placeholder-stt";

  async transcribe(audio: Buffer, mimeType: string): Promise<TranscriptResult> {
    const secondsEstimate = Math.max(1, Math.round(audio.length / 16_000));
    return {
      ok: true,
      text: `(stt placeholder · ~${secondsEstimate}s of ${mimeType} captured)`,
      provider: this.name,
    };
  }
}

export class WhisperTranscriber implements Transcriber {
  readonly name = "openai-whisper";

  constructor(
    private readonly apiKey: string,
    private readonly baseUrl = optionalEnv("OPENAI_BASE_URL", "https://api.openai.com/v1"),
    private readonly model = optionalEnv("TOKYOX_STT_MODEL", "whisper-1")
  ) {}

  async transcribe(audio: Buffer, mimeType: string): Promise<TranscriptResult> {
    const ext = mimeType.includes("webm")
      ? "webm"
      : mimeType.includes("ogg")
        ? "ogg"
        : mimeType.includes("mp4")
          ? "mp4"
          : "wav";
    try {
      const form = new FormData();
      form.append("file", new Blob([new Uint8Array(audio)], { type: mimeType }), `audio.${ext}`);
      form.append("model", this.model);
      const res = await fetch(`${this.baseUrl}/audio/transcriptions`, {
        method: "POST",
        headers: { authorization: `Bearer ${this.apiKey}` },
        body: form,
        signal: AbortSignal.timeout(30_000),
      });
      if (!res.ok) {
        return {
          ok: false,
          text: "",
          provider: this.name,
          reason: `HTTP ${res.status}`,
        };
      }
      const data = (await res.json()) as { text?: string };
      return { ok: true, text: data.text ?? "", provider: this.name };
    } catch (err) {
      return {
        ok: false,
        text: "",
        provider: this.name,
        reason: err instanceof Error ? err.message : String(err),
      };
    }
  }
}

export function createTranscriber(): Transcriber {
  const key = providerKey("openai");
  if (key) return new WhisperTranscriber(key);
  return new PlaceholderTranscriber();
}

File: tokyox/voice/stt.py
from __future__ import annotations
import os
from abc import ABC, abstractmethod
from dataclasses import dataclass

import httpx


@dataclass
class TranscriptResult:
    ok: bool
    text: str
    provider: str
    reason: str | None = None


class Transcriber(ABC):
    @property
    @abstractmethod
    def name(self) -> str: ...

    @abstractmethod
    async def transcribe(self, audio: bytes, mime_type: str) -> TranscriptResult: ...


class PlaceholderTranscriber(Transcriber):
    name = "placeholder-stt"

    async def transcribe(self, audio: bytes, mime_type: str) -> TranscriptResult:
        secs = max(1, len(audio) // 16000)
        return TranscriptResult(
            ok=True,
            text=f"(stt placeholder · ~{secs}s of {mime_type} captured)",
            provider=self.name,
        )


class WhisperTranscriber(Transcriber):
    name = "openai-whisper"

    def __init__(self, api_key: str, base_url: str | None = None, model: str | None = None):
        self._api_key = api_key
        self._base_url = base_url or os.environ.get("OPENAI_BASE_URL", "https://api.openai.com/v1")
        self._model = model or os.environ.get("TOKYOX_STT_MODEL", "whisper-1")

    async def transcribe(self, audio: bytes, mime_type: str) -> TranscriptResult:
        ext = "webm" if "webm" in mime_type else "ogg" if "ogg" in mime_type else "mp4" if "mp4" in mime_type else "wav"
        try:
            files = {"file": (f"audio.{ext}", audio, mime_type)}
            data = {"model": self._model}
            async with httpx.AsyncClient(timeout=30.0) as client:
                resp = await client.post(
                    f"{self._base_url}/audio/transcriptions",
                    headers={"Authorization": f"Bearer {self._api_key}"},
                    files=files,
                    data=data,
                )
                if not resp.is_success:
                    return TranscriptResult(ok=False, text="", provider=self.name, reason=f"HTTP {resp.status_code}")
                data = resp.json()
                return TranscriptResult(ok=True, text=data.get("text", ""), provider=self.name)
        except Exception as err:
            return TranscriptResult(ok=False, text="", provider=self.name, reason=str(err))


def create_transcriber() -> Transcriber:
    key = os.environ.get("OPENAI_API_KEY")
    if key:
        return WhisperTranscriber(key)
    return PlaceholderTranscriber()